### ISSUE TITLE
Add ccls/cquery (language server) to c-c++ layer

### ISSUE THREAD

**Issue by MaskRay** [syl20bnr/spacemacs#10134]
tl;dr Jump to the bottom to see the demo.

"cquery is a highly-scalable, low-latency language server for C/C++/Objective-C". There are uses on indexing Chromium, LLVM projects, and Linux kernel. It provides similar features with venerable rtags, like finding definitions/references.

https://github.com/MaskRay/ccls libclang + Clang C++
https://github.com/cquery-project/cquery libclang

You can take this issue as a "seeking for help". There are many pieces need customizing ([cquery.el](https://github.com/jacobdufault/cquery/blob/master/emacs/cquery.el) [lsp-mode](https://github.com/emacs-lsp/lsp-mode) [lsp-ui](https://github.com/emacs-lsp/lsp-ui)) to provide a kinda satisfying code assistant UI. spacemacs is used by many and to me a suitable place to combine these efforts.  I leave some notes at https://github.com/jacobdufault/cquery/wiki/Emacs but my elisp-fu is incapable to create a robust layer.

I'd also like to know if there is interest to **incorporate lsp-mode through which we can provide a unified interface for C++, Rust and many other languages**. I have used rls (Rust) and cquery (C++) and many features can and should be shared and be provided by a common library (lsp-mode lsp-ui). [cquery.el](https://github.com/jacobdufault/cquery/blob/master/emacs/cquery.el) provides a LSP plugin for cquery and it also has some extensions not included in LSP:

```elisp
;; emacs-ccls https://melpa.org/#/ccls
(ccls-xref-find-custom "$ccls/base")
(ccls-xref-find-custom "$ccls/callers")
;; No "$ccls/derived", use lsp-goto-implementation from  lsp-mode instead
(ccls-xref-find-custom "$ccls/vars")

;; Alternatively, use lsp-ui-peek interface
(lsp-ui-peek-find-custom 'base "$ccls/base")
(lsp-ui-peek-find-custom 'callers "$ccls/callers")

;; ...... emacs-cquery https://melpa.org/#/cquery
;; the cquery counterparts are similar
(cquery-xref-find-custom "$cquery/base")
(cquery-xref-find-custom "$cquery/callers")
(cquery-xref-find-custom "$cquery/derived")
(cquery-xref-find-custom "$cquery/vars")

;; Alternatively, use lsp-ui-peek interface
(lsp-ui-peek-find-custom 'base "$cquery/base")
(lsp-ui-peek-find-custom 'callers "$cquery/callers")
;; ......
```

There are still useful features that are missing or can be improved in [cquery.el](https://github.com/jacobdufault/cquery/blob/master/emacs/cquery.el). And it is also nice to learn from users using other code assistants to know what can be improved generally about cquery.

I might be biased regarding rtags because I am a contributor of cquery. What bothered me a lot when I was using rtags was its query performance https://github.com/Andersbakken/rtags/issues/1007

See also:
cquery.el: https://github.com/cquery-project/emacs-cquery/blob/master/cquery.el
lsp-mode https://github.com/emacs-lsp/lsp-mode
lsp-ui https://github.com/emacs-lsp/lsp-ui

A brief list of what can be improved

* semantic highlighting https://github.com/jacobdufault/cquery/issues/239
* Visualization of `$cquery/typeHierarchyTree`
* With `(setq cquery-extra-init-params '(:enableComments 2 :cacheFormat "msgpack")`, comments are not nicely rendered in eldoc   https://github.com/emacs-lsp/lsp-mode/pull/224
* https://github.com/jacobdufault/cquery/issues/30  some entries lean to VSCode, but many apply also to Emacs
* Some cases where libclang is incapable to provide all desired information https://github.com/jacobdufault/cquery/issues/252
* Something possible technically but entails a complicated implementation https://github.com/jacobdufault/cquery/issues/260
* Niche https://github.com/jacobdufault/cquery/issues/85 https://github.com/jacobdufault/cquery/issues/283

![rainbow semantic highlighting](https://ptpb.pw/rO3t.jpg)
![comments](https://camo.githubusercontent.com/fe1e12f9be72c2295d732d6265b42bde0d121ee8/68747470733a2f2f707470622e70772f5a6275462e6a7067)
![semantic highlighting](https://user-images.githubusercontent.com/320163/34711725-5b7b399c-f55b-11e7-93dc-0631aa154ce7.png)
![completion](https://ptpb.pw/lDaw.jpg)
![approximate workspace/symbol search](https://ptpb.pw/KOKn.jpg)

If you speak Chinese, I've written an article introducing cquery and my configuration http://maskray.me/blog/2017-12-03-c++-language-server-cquery

**Comment by MaskRay** [comment]
Adding a bunch of people who may be interested in Emacs

@scturtle @jiegec @topisani @danielmartin

**Comment by danielmartin** [comment]
I don't particularly use Spacemacs, only vanilla GNU Emacs, so I don't know what the requirements are for including a package in a particular layer. I think cquery.el is already pretty stable and provides a lot of functionality so it can be a good candidate for inclusion. Also, with cquery everything is in one place, and it implements a well-known protocol. Without LSP it is typical to have to install several specialized packages to have an IDE experience (rtags for code navigation and irony-mode/ycm for code completion, for example).

**Comment by sebastiencs** [comment]
I'm not using spacemacs either but if you're willing to integrate [lsp-ui](https://github.com/emacs-lsp/lsp-ui) I can help :)
An integration with lsp-rust would be nice too @bkchr

**Comment by bkchr** [comment]
I currently maintain my rust language server layer here: https://github.com/bkchr/rustrls
It would be nice if the interface could be unified across languages, but I don't have any idea up to now.

**Comment by scturtle** [comment]
If you want to have a try, I have a personal [lsp-cquery](https://github.com/scturtle/dotfiles/blob/master/spacemacs/.emacs.d/private/lsp-cquery/packages.el) layer with some [customizations](https://github.com/scturtle/dotfiles/blob/0c0765f7fa331f652a6a4188427ecb78a1540ee0/spacemacs/.spacemacs#L442-L454).

Agree with @bkchr. If there is a lsp layer with common configurations, it would be easier to integrate lsp client into each language layer.

**Comment by MaskRay** [comment]
I have also cleaned my lsp-mode/lsp-ui based [config](https://github.com/MaskRay/Config/blob/master/home/.emacs.d/private/%2Bmy/my-code/packages.el). The [cquery specific stuff](https://github.com/MaskRay/Config/blob/master/home/.emacs.d/private/%2Bmy/my-cquery/packages.el) is very simple now.

**Comment by MaskRay** [comment]
Created https://github.com/syl20bnr/spacemacs/pull/10211 and [emacs-cquery is now in Melpa(https://github.com/melpa/melpa/pull/5235)

@syl20bnr

**Comment by cormacc** [comment]
@MaskRay I spent a few hours on a train journey today comparing your cquery and my-code layers with the existing c-c++ layer, paying attention to how ycmd and rtags are integrated in the latter. Have refactored some of the functionality into a reasonably coherent set of additions to the lsp layer with some of the common keybindings, a new cquery layer and some minimal additions to the c-c++ layer in a new branch forked from spacemacs/develop.

Think it's fairly close, but certainly not suitable for a a single PR -- not sure if it's of any use to you, or where to go with it from here, but it's here;
https://github.com/cormacc/spacemacs/tree/cquery-layer

**Comment by MaskRay** [comment]
Awesome work! I left some notes.

**Comment by myrgy** [comment]
> @cormacc
> Reworked some of MaskRay's config into a a new PR for the lsp layer, here:
> https://github.com/syl20bnr/spacemacs/pull/10486
> 
> And an attempt at a cquery layer here:
> cormacc/spacemacs@b8ae1bc
> 
> I'll raise a new PR for the cquery layer if the lsp-layer one gets merged (it builds on it).

**Comment by arefasvadi** [comment]
@cormacc 
I am not an elisp or spacemacs guru, but I think your solution is closest to what I want (cmake support, auto-completion,...). I previously managed to get rtags working, but it's too slow! 
I have cloned the "[new-cquery-layer](https://github.com/cormacc/spacemacs.git)" branch from your forked repo. Can I ask your configuraions for .spacemacs file!
Currently I have enabled the following layers:
1. cquery
2. cmake with cmake-enable-cmake-ide-support
3. c++ with variables for cmake-ide-support, enable-clang-support, enable-c++11 support.
4. lsp

**Comment by cormacc** [comment]
@arefasvadi 
Simplest thing is to just include cquery layer. It should include c-c±± and lsp layers automatically -- they're declared as dependencies in the cquery layers.el

Here's my config -- I include lsp layer explicitly to override some defaults...

I enable cmake layer as well, but just for CMAKELISTS syntax support. i.e. without enabling cmake-ide support, as that has some functionality that overlaps/conflicts.

Though it would be nice at some stage to use cmake-ide to pick up changes to cmakelists.txt and regenerate compile_commands.json, using a persistent build pool directory, and configure query to look there rather than in the project root. Assuming you can disable the rest of the cmake-ide functionality (rtags etc.). Not useful enough for any of my current projects to tackle it just now though --- sounds like it would take a while to figure out :) I just regenerate compile_commands.json manually every now and then.

```
    dotspacemacs-configuration-layers
    '(
....
       ;; lsp  and c-c++ layers implicitly loaded by cquery layer...
       (lsp :variables
         ;; ... though loading lsp layer explicitly here to override some defaults
         lsp-ui-peek-expand-by-default t)
       cquery
       cmake
....
       )
```

**Comment by MaskRay** [comment]
Time passes, I no longer use cquery and have a different plan. I made a fork on Apr 1 (sarcastically!) and renovated it with improvement of import pipeline last weekend and clang+llvm C++ API this weekend. It has only 3 git submodules and the removal of loguru is also in the works. If you are curious enough and intend to give it a try:

https://github.com/MaskRay/ccls/wiki/Getting-started

Caution: C++17 and Clang C++ API are tough to tame on old distributions.

**Comment by sdwolfz** [comment]
Yes, thanks everybody for participating in this. It was a huge piece of work ❤️!
